File: crazyflie_lab/crazyflie_amr_cpp/src/pid_tuner_answer.cpp
#include <rclcpp/rclcpp.hpp>
#include <rmw/qos_profiles.h>

#include <rcl_interfaces/srv/set_parameters.hpp>
#include <rcl_interfaces/msg/parameter.hpp>
#include <geometry_msgs/msg/twist.hpp>
#include <crazyflie_interfaces/msg/log_data_generic.hpp>

#include <chrono>
#include <algorithm>

using namespace std::chrono_literals;

class PidTunerNode : public rclcpp::Node
{
public:
  ~PidTunerNode(void) = default;

  PidTunerNode(const std::string &cf_name="cf80") 
    : Node("pid_tuner"),
    _cf_name(cf_name),
    _vel_cmd_active(false),
    _ang_vel_received(false)
  {
    auto qos_profile = rclcpp::QoS(rclcpp::QoSInitialization::from_rmw(rmw_qos_profile_sensor_data), rmw_qos_profile_sensor_data);

    _vel_sub = this->create_subscription<geometry_msgs::msg::Twist>(
      "/" + cf_name + "/cmd_vel_legacy", qos_profile,
      [this](const geometry_msgs::msg::Twist &msg) { this->vel_callback(msg); });

    _data_sub = this->create_subscription<crazyflie_interfaces::msg::LogDataGeneric>(
      "/" + cf_name + "/gyro_data", qos_profile,
      [this](const crazyflie_interfaces::msg::LogDataGeneric &msg) { this->data_callback(msg); });

    _setup_timer = this->create_wall_timer(500ms, [this](void) { this->setup(); });

    _set_param_client = this->create_client<rcl_interfaces::srv::SetParameters>("/crazyflie_server/set_parameters");
  }

private:

  void setup(void) 
  {
    if(_set_param_client->service_is_ready()) {
      RCLCPP_INFO(rclcpp::get_logger("pid_tuner"), "Set Parameter Service is ready.");

      _pitch_Kp = kStartKp;
      _pitch_Ki = kStartKi;
      _pitch_Kd = kStartKd;

      _stable_1  = false;
      _stable_2  = false;
      _oscillate = false;

      rcl_interfaces::msg::Parameter kp;
      kp.name               = _cf_name + ".params.pid_rate.pitch_kp";
      kp.value.type         = rcl_interfaces::msg::ParameterType::PARAMETER_DOUBLE;
      kp.value.double_value = _pitch_Kp;

      rcl_interfaces::msg::Parameter ki;
      ki.name               = _cf_name + ".params.pid_rate.pitch_ki";
      ki.value.type         = rcl_interfaces::msg::ParameterType::PARAMETER_DOUBLE;
      ki.value.double_value = _pitch_Ki;

      rcl_interfaces::msg::Parameter kd;
      kd.name               = _cf_name + ".params.pid_rate.pitch_kd";
      kd.value.type         = rcl_interfaces::msg::ParameterType::PARAMETER_DOUBLE;
      kd.value.double_value = _pitch_Kd;

      _set_param_req = std::make_shared<rcl_interfaces::srv::SetParameters::Request>();
      _set_param_req->parameters = std::vector<rcl_interfaces::msg::Parameter>({kp, ki, kd});
      _set_param_result = _set_param_client->async_send_request(_set_param_req).share();

      _cmd_timer = this->create_wall_timer(1000ms, [this](void) { this->control_func(); });
      _setup_timer->cancel();
    } else {
      RCLCPP_WARN(rclcpp::get_logger("pid_tuner"), "Set Parameter Service is NOT ready.");
    }
  }

  void control_func(void)
  {
    if(_ang_vel_received && _vel_cmd_active) {
      
      if(_set_param_result.valid()) {
        if(_set_param_result.wait_for(std::chrono::seconds(0)) == std::future_status::ready) {
          _set_param_resp = _set_param_result.get();
          std::vector<rcl_interfaces::msg::SetParametersResult> results =  _set_param_resp->results;

          // Reset the future so that valid will return false
          _set_param_result = rclcpp::Client<rcl_interfaces::srv::SetParameters>::SharedFuture();
        }
      }

      // If call has finished, update gains and set parameters again
      if(!_set_param_result.valid()) {

        // Standard deviation of data
        std::vector<double> stdev;
        stdev.push_back(stdev_compute(_ang_data_x));
        stdev.push_back(stdev_compute(_ang_data_y));
        stdev.push_back(stdev_compute(_ang_data_y));

        // Min/Max standard deviation of data
        auto minmax_stdev = std::minmax_element(stdev.begin(), stdev.end());
        double min_stdev = *minmax_stdev.first;
        double max_stdev = *minmax_stdev.second;

        RCLCPP_INFO(rclcpp::get_logger("pid_tuner"), std::string("Min Stdev: " + std::to_string(min_stdev)).c_str()  );
        RCLCPP_INFO(rclcpp::get_logger("pid_tuner"), std::string("Max Stdev: " + std::to_string(max_stdev)).c_str()  );

        // Determine if UAV has switched from statble to oscillating to back to stable
        if ((!_stable_1) && (min_stdev < 5.0) && (max_stdev < 10.0)) {
          RCLCPP_INFO(rclcpp::get_logger("pid_tuner"), "Stablized: First");
          _stable_1 = true;
        } else if (_stable_1 && (max_stdev > 65.0)) {
          RCLCPP_INFO(rclcpp::get_logger("pid_tuner"), "Oscillating");
          _oscillate = true;
        } else if (_stable_1 && _oscillate && (min_stdev < 5.0) && (max_stdev < 10.0)) {
          RCLCPP_INFO(rclcpp::get_logger("pid_tuner"), "Crazyflie has Stabilized!");
          _stable_2 = true;
        }

        // Adjust gains if necessary
        if ((!_stable_1) || (!_oscillate)) {
          _pitch_Kp += 20.0; //20.0;
        } else if (_stable_1 && _oscillate && (!_stable_2)) {
          _pitch_Kd += 0.2; //0.2;
        }

        // Call to set parameters service
        _set_param_req->parameters[0].value.double_value = _pitch_Kp;
        _set_param_req->parameters[1].value.double_value = _pitch_Ki;
        _set_param_req->parameters[2].value.double_value = _pitch_Kd;

        _set_param_result = _set_param_client->async_send_request(_set_param_req).share();
      }

      _ang_vel_received = false;
    }
  }

  void data_callback(const crazyflie_interfaces::msg::LogDataGeneric &msg)
  {
    _ang_data_x.push_back(msg.values[0]);
    _ang_data_y.push_back(msg.values[1]);
    _ang_data_z.push_back(msg.values[2]);

    if(_ang_data_x.size() > kMaxDataLength) {
      _ang_data_x.erase(_ang_data_x.begin());
      _ang_data_y.erase(_ang_data_y.begin());
      _ang_data_z.erase(_ang_data_z.begin());

      _ang_vel_received = true;
    }
  }

  void vel_callback(const geometry_msgs::msg::Twist &msg) 
  {
    _vel_cmd_active = msg.linear.z > 0.0;
  }

  double stdev_compute(const std::vector<double> &data)
  {
    // Number of elements
    size_t size = data.size();

    // Calculate the sum of elements in the vector
    double mean = std::accumulate(data.begin(), data.end(), 0.0) / size;

    // Calculate the sum of squared differences from the mean
    double stdev = 0.0;
    for (int i = 0; i < size; ++i) {
        stdev += pow(data[i] - mean, 2);
    }

    // Calculate the square root of the variance
    return sqrt(stdev / size);
  }

  // ---------------------------------//
  // Variable used by the Node Object //
  // ---------------------------------//

  rclcpp::Subscription<geometry_msgs::msg::Twist>::SharedPtr _vel_sub;
  rclcpp::Subscription<crazyflie_interfaces::msg::LogDataGeneric>::SharedPtr _data_sub;
  rclcpp::TimerBase::SharedPtr _cmd_timer;
  rclcpp::TimerBase::SharedPtr _setup_timer;

  rclcpp::Client<rcl_interfaces::srv::SetParameters>::SharedPtr    _set_param_client;
  rclcpp::Client<rcl_interfaces::srv::SetParameters>::SharedFuture _set_param_result;
  std::shared_ptr<rcl_interfaces::srv::SetParameters::Request>     _set_param_req;
  std::shared_ptr<rcl_interfaces::srv::SetParameters::Response>    _set_param_resp;

  std::string _cf_name;

  bool _vel_cmd_active;
  bool _ang_vel_received;
  std::vector<double> _ang_data_x;
  std::vector<double> _ang_data_y;
  std::vector<double> _ang_data_z;

  double _pitch_Kp;
  double _pitch_Ki;
  double _pitch_Kd;

  bool _stable_1;
  bool _stable_2;
  bool _oscillate;

  static constexpr double kStartKp = 10.0;
  static constexpr double kStartKi = 0.0;
  static constexpr double kStartKd = 0.0;

  static constexpr size_t kMaxDataLength = 5;
};


int main(int argc, char *argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<PidTunerNode>());
  rclcpp::shutdown();

  return 0;
}

File: crazyflie_lab/crazyflie_amr_py/crazyflie_amr/pid_tuner_answer.py
#!/usr/bin/env python

import numpy as np

import rclpy
from rclpy.node import Node
from rclpy.task import Future
from rclpy.qos import QoSProfile, QoSReliabilityPolicy
from geometry_msgs.msg import Twist

from crazyflie_interfaces.srv import *
from crazyflie_interfaces.msg import LogDataGeneric

from rcl_interfaces.srv import *
from rcl_interfaces.msg import *


START_Kp = 0.0
START_Ki = 0.0
START_Kd = 0.0

class PIDTunerNode(Node):
    def __init__(self, cf_name = 'cf80'):
        super().__init__('pid_tuner')
        self.cf = cf_name

        qos_profile = QoSProfile(depth=10, reliability=QoSReliabilityPolicy.BEST_EFFORT)
        self.gyro_sub = self.create_subscription(LogDataGeneric, f"/{self.cf}/gyro_data", self.gyro_callback, qos_profile)
        self.vel_sub  = self.create_subscription(Twist, f"/{self.cf}/cmd_vel_legacy", self.vel_callback, qos_profile)
        
        self.timer_period_s = 1.0
        self.setup_timer = self.create_timer(self.timer_period_s, self.setup)
        self.cmd_timer   = None

        # Variables for Node
        self.pitch_kp = START_Kp
        self.pitch_ki = START_Ki
        self.pitch_kd = START_Kd

        self.ang_vel_received = False
        self.ang_data_x = []
        self.ang_data_y = []
        self.ang_data_z = []

        self.stable_1  = False
        self.stable_2  = False
        self.oscillate = False

        self.vel_active       = False
        self.data_initialized = False

        # Clients to get/set parameters
        param_names = [
            f"{self.cf}.params.pid_rate.pitch_kp", 
            f"{self.cf}.params.pid_rate.pitch_ki", 
            f"{self.cf}.params.pid_rate.pitch_kd"]

        self.kp_idx = 0
        self.ki_idx = 1
        self.kd_idx = 2

        self.set_client       = self.create_client(SetParameters, '/crazyflie_server/set_parameters')
        self.set_msg          = SetParameters.Request()
        self.param_set_params = [Parameter(name=nm, value=ParameterValue(type=ParameterType.PARAMETER_DOUBLE)) for nm in param_names]
        self.set_future       = None
        

    def setup(self):
        if self.set_client.service_is_ready():
            self.get_logger().warn("Parameter service ready!")

            self.param_set_params[self.kp_idx].value.double_value = self.pitch_kp
            self.param_set_params[self.ki_idx].value.double_value = self.pitch_ki
            self.param_set_params[self.kd_idx].value.double_value = self.pitch_kd

            self.set_msg.parameters = self.param_set_params
            self.set_future         = self.set_client.call_async(self.set_msg)

            self.setup_timer.cancel()
            self.cmd_timer   = self.create_timer(self.timer_period_s, self.control_func)
        else:
            self.get_logger().info("Parameter service NOT ready")

    def control_func(self):

        if self.ang_vel_received and self.vel_active:
            self.ang_vel_received = False

            # First, checking if call to set parameter has finished
            if self.set_future:
                if self.set_future.done():
                    response = self.set_future.result()
                    if response is not None:
                        for rst in response.results:
                            if not rst.successful:
                                self.get_logger().warn(f"Failed to set parameter. Reason: {rst.reason}")
                    else:
                        self.get_logger().warn("Call to set parameter service failed")
                    
                    self.set_future = None
                    
            # If call has finished, future will be set to "None"
            # Update gains and set parameters again
            if not self.set_future:

                ang_x_std = np.std(self.ang_data_x)
                ang_y_std = np.std(self.ang_data_y)
                ang_z_std = np.std(self.ang_data_z)

                min_std = np.min([ang_x_std, ang_y_std, ang_z_std])
                max_std = np.max([ang_x_std, ang_y_std, ang_z_std])

                self.get_logger().info(f"Min std: {min_std:0.2f}")
                self.get_logger().info(f"Max std: {max_std:0.2f}")

                # Determine if UAV has switched from statble to oscillating to back to stable
                if (not self.stable_1) and min_std < 5.0 and max_std < 10.0:
                    self.stable_1 = True
                    self.get_logger().info(f"Stablized 1")
                elif self.stable_1 and max_std > 65.0:
                    self.oscillate = True
                    self.get_logger().info(f"Oscillating")
                elif self.stable_1 and self.oscillate and min_std < 5.0 and max_std < 10.0:
                    self.stable_2 = True
                    self.get_logger().info(f"Crazyflie has Stabilized!")

                # Adjust gains to set
                if (not self.stable_1) or (not self.oscillate):
                    self.pitch_kp += 0.0 #20.0
                elif self.stable_1 and self.oscillate and (not self.stable_2):
                    self.pitch_kd += 0.0 #0.2

                # Call to service to set parameters
                self.param_set_params[self.kp_idx].value.double_value = self.pitch_kp
                self.param_set_params[self.ki_idx].value.double_value = self.pitch_ki
                self.param_set_params[self.kd_idx].value.double_value = self.pitch_kd

                self.set_msg.parameters = self.param_set_params
                self.set_future         = self.set_client.call_async(self.set_msg)
 

    def gyro_callback(self, msg):

        self.ang_data_x.append(msg.values[0])
        self.ang_data_y.append(msg.values[1])
        self.ang_data_z.append(msg.values[2])

        if len(self.ang_data_x) > 5:
            self.ang_data_x = self.ang_data_x[1:]
            self.ang_data_y = self.ang_data_y[1:]
            self.ang_data_z = self.ang_data_z[1:]
            
            self.ang_vel_received = True

    def vel_callback(self, msg):
        self.vel_active = msg.linear.z > 0.0
        
def main(args=None):
    rclpy.init(args=args)
    
    pid_tune_node = PIDTunerNode()
    rclpy.spin(pid_tune_node)

    pid_tune_node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
